test/jdk/tools/jlink/JLink20000Packages.java

```diff
@@ -0,0 +1,129 @@
+ *          pagination is in place, the limitation is on the constant pool size, not number
+ *          of packages.
+ * @bug 8321413
+ * @library ../lib
+ * @enablePreview
+ * @modules java.base/jdk.internal.jimage
+ *          jdk.jlink/jdk.tools.jlink.internal
+ *          jdk.jlink/jdk.tools.jlink.plugin
+ *          jdk.jlink/jdk.tools.jmod
+ *          jdk.jlink/jdk.tools.jimage
+ *          jdk.compiler
+ * @build tests.*
+ * @run main/othervm -Xmx1g -Xlog:init=debug -XX:+UnlockDiagnosticVMOptions -XX:+BytecodeVerificationLocal JLink20000Packages
+ */
+public class JLink20000Packages {
+    private static final ToolProvider JAVAC_TOOL = ToolProvider.findFirst("javac")
+            .orElseThrow(() -> new RuntimeException("javac tool not found"));
+
+    static void report(String command, String[] args) {
+        System.out.println(command + " " + String.join(" ", Arrays.asList(args)));
+    }
+
+    static void javac(String[] args) {
+        report("javac", args);
+        JAVAC_TOOL.run(System.out, System.err, args);
+    }
+
+    public static void main(String[] args) throws Exception {
+        Path src = Paths.get("bug8321413");
+        Path imageDir = src.resolve("out-jlink");
+        Path mainModulePath = src.resolve("bug8321413x");
+
+        StringJoiner mainModuleInfoContent = new StringJoiner(";\n  exports ", "module bug8321413x {\n  exports ", ";\n}");
+
+        for (int i = 0; i < 20000; i++) {
+            String packageName = "p" + i;
+            String className = "C" + i;
+
+            Path packagePath = Files.createDirectories(mainModulePath.resolve(packageName));
+
+            StringBuilder classContent = new StringBuilder("package ");
+            classContent.append(packageName).append(";\n");
+            classContent.append("class ").append(className).append(" {}\n");
+            Files.writeString(packagePath.resolve(className + ".java"), classContent.toString());
+
+            mainModuleInfoContent.add(packageName);
+        }
+
+        // create module reading the generated modules
+        Path mainModuleInfo = mainModulePath.resolve("module-info.java");
+        Files.writeString(mainModuleInfo, mainModuleInfoContent.toString());
+
+        Path mainClassDir = mainModulePath.resolve("testpackage");
+        Files.createDirectories(mainClassDir);
+
+        Files.writeString(mainClassDir.resolve("JLink20000PackagesTest.java"), """
+                package testpackage;
+
+                public class JLink20000PackagesTest {
+                    public static void main(String[] args) throws Exception {
+                        System.out.println("JLink20000PackagesTest started.");
+                    }
+                }
+                """);
+
+        String out = src.resolve("out").toString();
+        javac(new String[]{
+                "-d", out,
+                "--module-source-path", src.toString(),
+                "--module", "bug8321413x"
+        });
```

Comment:
FYI.   You can consider using `jdk.test.lib.compiler.CompilerUtils::compile` method to invoke javac in-process.